### ISSUE TITLE
might be better to check what packages get required during tests/build

### ISSUE THREAD

**Issue by jedwards1211** [depcheck/depcheck#192]
Considering the numerous dynamic ways in which packages can be required, I think it will always be hard to statically analyze which packages are used.

Here's a more foolproof suggestion: run the tests and build with a require hook (which looks like [this](https://github.com/mdlawson/piping/blob/master/src/launcher.coffee#L13)) that keeps track of what actually gets required, and then check that against `package.json`.

That way no matter how a package gets required, whether as a webpack loader, `mocha -r jsdom/global` etc., `depcheck` will know about it.

**Comment by lijunle** [comment]
I think this is an awesome idea! I love it! I think it is worth to open another project for this and not affecting the current logic.

My understanding is like this:

- In the beginning of the test code, `require('depcheck/inspect')`.
- It hook on `require` and `process.exit` function to collect/write data.
- There is another CLI tool to generate report from the data (or may we pollute the test output).

How do you think about this?

**Comment by jedwards1211** [comment]
Right, actually I'm thinking it's ideal for it to be a command runner like istanbul's `nyc`.  I may try to start a project that does this soon.

**Comment by lijunle** [comment]
My concern is, there is so much prefix command. If it is named `depcoverage`, the command will go to `depcoverage nyc mocha` which looks ugly to me. 🐶 

I could suggest to be a register module, which can be used by `nyc` config or `mocha -r` to include this module. Then, we have another step to collect the data, like `nyc --report`.

Of course, we could have both.

I would like to help on the project. Do you want to contribute the project to the org? I am sending the invitation to you. 😸

**Comment by jedwards1211** [comment]
Agreed, it's a lot of nesting.  I'm not sure if it will work with `mocha -r` because I think one might have to patch `module._load` and then `require` the code to be run.  But I'll have to find out.

My boss may want me to put the project under our company's org, for marketing's sake, but of course I got half of the idea from depcheck anyway, so I'll try to convince him to let me contribute the project to you guys.

**Comment by jedwards1211** [comment]
Ah, so `_load` is actually a global in `require("module")._load`.  So we could use `mocha -r`.  

However, like `nyc`, we would want to do something (checking deps) after the tests run, and I think the way `nyc` manages that is having the child process it runs your tests in (and **its* descendant processes too!) send their coverage back to `nyc` in a `process.on('exit')` hook.  This way no matter how many subprocesses your tests spawn, `nyc` can catch 'em all (and I have written integration tests that need to capture test coverage from subprocesses).  I still need to do a bit of research on how exactly `nyc` works.

And if we patch our hook inside the process spawned by `nyc`, then we may not find out that the `nyc` package is being used.  So I think making `depcoverage` run as the outermost process is the only way it will work.

I just realized though, there may be packages the tests don't use that are needed for other commands, like `semantic-release`, `@kadira/storybook`, etc...so it's not exactly an easy problem.

**Comment by lijunle** [comment]
@jedwards1211 Under your company's org is fine. I care more about the license. If that is friendly, I can contribute to it and leverage it back in depcheck. 

As you mentioned, I am not sure how much it can help. For some cases, like `eslint`, a simple static analysis + package.json discover looks a better choice than runtime hook. There are many other tools is not a part of test - babel, webpack, gulp, etc.

A crazy idea is to prefix `depcoverage` before all these commands. I am not sure how well it works.

Maybe we could one step further, if a flag is set, **all** nodejs commands run under a require hook. After all runs, collect the data to generate report. It sounds like deeper than `module._load`. I am not sure if nodejs can do so.

**Comment by jedwards1211** [comment]
@lijunle well, I typically run `eslint`, `flow`, test, and build in my `prepublish` scripts, so theoretically I just need to run `depcoverage npm prepublish`.

**Comment by lijunle** [comment]
I read the code of `nyc`, they are not using `module._load`. It is require extension: https://gist.github.com/jamestalmage/df922691475cff66c7e6

**Comment by lijunle** [comment]
Take a deep look into the `spawn-wrap` package. Basically, it hooks on [`ChildProcess#spawn` and `spawn_sync.spawn`](https://github.com/tapjs/spawn-wrap/blob/master/index.js#L64-L67) functions. It use various ways to [detect it](https://github.com/tapjs/spawn-wrap/blob/master/index.js#L82) is calling to a node/iojs call, then redirect the call to a [prepared shim script](https://github.com/tapjs/spawn-wrap/blob/master/index.js#L298).

In our case, I am thinking about, because in CI build, the node binary is [configured by `nvm`](https://travis-ci.org/#L252). We might use a similar way to hook the node binary to our customized one. However, my first concern is the performance as mentioned in [spawn-wrap caveats](https://github.com/tapjs/spawn-wrap#contracts-and-caveats).

**Comment by jedwards1211** [comment]
wow, it's unfortunate that spawn_sync has to be so complicated!  That's neat though, you've looked into this deeper than I have already.

What they say about performance is kind of vague though.  I would think they mean it just slows down the startup of spawned processes?  Certainly I wish mocha + babel + nyc would start faster (it takes 5 seconds or so on my machine) but once it's up the tests seem to run at normal speed.

Also I'm not sure if nyc has to do anything special for CI builds, but I've never had problems using it in Travis CI.

**Comment by lijunle** [comment]
I understand *depcheck* is pretty slow because of its bad IO bomb in one thread. I want to have numbers here to exactly identify the problem and improve it. It is lesson for next project to think about metric from the beginning.

In regards to the CI build, I don't mean *nyc* is doing some special in CI environment. I am still thinking about how the end experience will be. 

- Option 1

  ```sh
  depcoverage nyc mocha ...
  ```

- Option 2

  ```sh
  depcoverage start
  nyc mocha ...
  eslint ...
  depcoverage end
  ```

- Option 3

  ```sh
  depcoverage nyc mocha ...
  depcoverage eslint ...
  depcoverage report // not report until explicitly report
  ```

From the foresee limitation and compatibility, I could suggest *option 3* here. What is your idea?

**Comment by jedwards1211** [comment]
Weird, I thought I had replied to this.  I think I like Option 3 best.  It may be tricky to cover all modules for some projects, for instance for some of my webapps there are deps I only use for starting my app in dev mode, and it's not always worth the time to write integration tests for my dev mode scripts.

I have so much on my plate that I don't want to spend time implementing this right now, unfortunately.  But if you want to take a stab at it, feel free!  And I'm happy to keep discussing it in any case.
